dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java

```diff
@@ -79,15 +78,21 @@ public int addObject(Object newClass) {
         int identityHashCode = capHashCode(System.identityHashCode(newClass), hashBit);
         long i = getIdentityAddress(identityHashCode, memoryAddress);
         int currentIndex = unsafe.getInt(i);
-        if (currentIndex == 0) {
-            int idx = fallbackMapEnabled ?
-                    fallbackMap.get((Class<?>) newClass) :
-                    atomicIndex.incrementAndGet();
-            unsafe.putIntVolatile(null, i, idx);
-            controlMap.put(newClass, idx);
-            return idx;
+        if(currentIndex == 0) {
+            int idx = fallbackMapEnabled ? fallbackMap.get((Class<?>) newClass) : atomicIndex.incrementAndGet();
+            // use cas to check, is there an index to an existing class
+            if(unsafe.compareAndSwapInt(null, i, 0, idx)) {
+                cache.set(idx - 1, newClass);
```

Comment:
We cache currentIndex without -1, we don't care about wasting a position if it helps readability